domain/application/state/application_test.go

```diff
@@ -902,6 +902,81 @@ func (s *applicationStateSuite) TestStorageDefaults(c *gc.C) {
 	})
 }
 
+func (s *applicationStateSuite) TestCreateApplicationThenGetCharmIDByApplicationName(c *gc.C) {
+	origin := charm.CharmOrigin{
+		Source:   charm.LocalSource,
+		Revision: 42,
+	}
+
+	expectedMetadata := charm.Metadata{
+		Name:           "ubuntu",
+		Summary:        "summary",
+		Description:    "description",
+		Subordinate:    true,
+		RunAs:          charm.RunAsRoot,
+		MinJujuVersion: version.MustParse("4.0.0"),
+		Assumes:        []byte("null"),
+	}
+	expectedManifest := charm.Manifest{
+		Bases: []charm.Base{
+			{
+				Name: "ubuntu",
+				Channel: charm.Channel{
+					Track: "latest",
+					Risk:  charm.RiskEdge,
+				},
+				Architectures: []string{"amd64", "arm64"},
+			},
+		},
+	}
+	expectedActions := charm.Actions{
+		Actions: map[string]charm.Action{
+			"action1": {
+				Description:    "description",
+				Parallel:       true,
+				ExecutionGroup: "group",
+				Params:         []byte(`{}`),
+			},
+		},
+	}
+	expectedConfig := charm.Config{
+		Options: map[string]charm.Option{
+			"option1": {
+				Type:        "string",
+				Description: "description",
+				Default:     "default",
+			},
+		},
+	}
+	expectedLXDProfile := []byte("[{}]")
+
+	_, err := s.state.CreateApplication(context.Background(), "foo", application.AddApplicationArg{
+		Charm: charm.Charm{
+			Metadata:   expectedMetadata,
+			Manifest:   expectedManifest,
+			Actions:    expectedActions,
+			Config:     expectedConfig,
+			LXDProfile: expectedLXDProfile,
+		},
+		Origin: origin,
+		Channel: &application.Channel{
+			Track:  "track",
+			Risk:   "stable",
+			Branch: "branch",
+		},
+		Platform: application.Platform{
+			OSType:       charm.Ubuntu,
+			Architecture: charm.AMD64,
+			Channel:      "22.04",
+		},
+	})
+	c.Assert(err, jc.ErrorIsNil)
+
+	chID, err := s.state.GetCharmIDByApplicationName(context.Background(), "foo")
+	c.Assert(err, jc.ErrorIsNil)
+	c.Check(chID.Validate(), jc.ErrorIsNil)
+}
+
```

Comment:
Since the `GetCharmIdByApplicationName` includes the promise of an error as part of its public interface this should really also be tested to ensure that it does not get accidentally quietly changed in the future.

Reply:
Correct, good catch.